### PR TITLE
Feature/array description (#528)

### PR DESCRIPTION
**Pull request by klausschuch**
see #528 
Converted schema section 2.2 of the array FCP doc file to adoc.

The changes are based on the last FMI3.0 word file (https://github.com/modelica/fmi-design/blob/master/FMI_3.0/FunctionalMockupInterface_3.0.docx)

some open problems:
- xml examples are missing
- ModelStructure/NumberOfEventIndicators is missing (only mentioned in api section 2.1)
...

### CODE REVIEW

File: docs/2_2_common_schema.adoc [hunk 1]
```diff
@@ -619,15 +619,31 @@ It provides the static information of all exposed variables and is defined as:
...
-The `ModelVariables` element consists of an ordered set of `ScalarVariable` elements (see figure above).
-The first element has `index = 1`,
-the second `index=2`, etc.
-This `ScalarVariable` `index` is used in element `ModelStructure` to uniquely and efficiently refer to `ScalarVariable` definitions.
-A `ScalarVariable` represents a variable of primitive type, like a real or integer variable.
-For simplicity,
-only scalar variables are supported in the schema file in this version and structured entities (like arrays or records) have to be mapped to scalars.
-The schema definition is present in a separate file `fmi3ScalarVariable.xsd`.
-The attributes of `ScalarVariable` are:
+The `ModelVariables` element consists of an ordered set of `Variable` elements (see figure above).
```
Comment (t-sommer) [inline docs/2_2_common_schema.adoc]: When citing form the XML, shouldn't we keep the angle brackets (e.g. `<Variable>`) and use the plain word "variable" when we refer to the abstract idea of a variable?

File: docs/2_2_common_schema.adoc [hunk 2]
```diff
@@ -672,16 +688,29 @@ The algebraic relationship to the inputs is defined via the `dependencies` attri
 - `"local"`: Local variable that is calculated from other variables or is a continuous-time state (see <<ModelStructure>>).
 It is not allowed to use the variable value in another model or slave.
 
-- `"independent""`: The independent variable (usually `time`).
+- `"independent"`: The independent variable (usually `time`).
```
Comment (t-sommer) [inline docs/2_2_common_schema.adoc]: I suggest to only use `independent` (w/o double quotes) except you cite a complete XML tag.

File: docs/2_2_common_schema.adoc [hunk 3]
```diff
@@ -864,24 +899,26 @@ calculated
 
 _[Note: (1) If `causality = "independent"`, it is neither allowed to define a value for `initial` nor a value for start.
 (2) If `causality = "input"`, it is not allowed to define a value for `initial` and a value for start must be defined.
-(3) If \(C) and `initial = "exact"`, then the variable is explicitly defined by its start value in Initialization Mode (so directly after calling `fmi3ExitInitializationMode`, the value of the variable is either the start value stored in element `<ScalarVariable><XXX start=YYY/>` or the value provided by `fmi3SetXXX`, if this function was called on this variable).]_
+(3) If \(C) and `initial = "exact"`, then the variable is explicitly defined by its start value in Initialization Mode (so directly after calling `fmi3ExitInitializationMode`, the value of the variable is either the start value stored in element `<Variable><XXX start=YYY/>` or the value provided by `fmi3SetXXX`, if this function was called on this variable).]_
```
Comment (t-sommer) [inline docs/2_2_common_schema.adoc]: This should be `<Variable start="x"/>` (with double quotes because it cites an imaginary XML tag)

File: docs/2_2_common_schema.adoc [hunk 4]
```diff
@@ -934,7 +975,7 @@ _Discussion of the combinations that are [underline]#not allowed#_:
 |_The combinations `"constant" / "parameter"`, `"constant" / "calculatedParameter"` and `"constant" / "input"` do not make sense, since parameters and inputs are set from the environment, whereas a constant has always a value._
 
 ^|_[red]#(b)#_
-|_The combinations `"discrete / parameter"`, `"discrete / calculatedParameter"`, `"continuous / parameter"` and `"continuous" / "calculatedParameter"` do not make sense, since `causality = "parameter"` and `"calculatedParameter"` define variables that do not depend on time, whereas `"discrete"` and `"continuous"` define variables where the values can change during simulation._
+|_The combinations `"discrete / structuralParameter"`, `"discrete / parameter"`, `"discrete / calculatedParameter"`, `"continuous / structuralParameter", `"continuous / parameter"` and `"continuous" / "calculatedParameter"` do not make sense, since `causality = "structuralParameter"`, causality = "parameter"` and `"calculatedParameter"` define variables that do not depend on time, whereas `"discrete"` and `"continuous"` define variables where the values can change during simulation._
```
Comment (t-sommer) [inline docs/2_2_common_schema.adoc]: This should be reformulated in plain english (is or equals vs. =)

File: docs/2_2_common_schema.adoc [hunk 5]
```diff
@@ -1032,14 +1081,14 @@ Even if this Dirac impulse would be modeled correctly by the modeling environmen
...
 
 _FMI for Model Exchange:_ +
-_Therefore, "tuning a parameter" during simulation does not mean to "change the parameter online" during simulation.
+_Therefore, "tuning a (structural) parameter" during simulation does not mean to "change the parameter online" during simulation.
 Instead, this is a short hand notation for:_
 
 . _Stop the simulation at an event instant (usually, a step event, in other words, after a successful integration step)._
 
-. _Change the values of the tunable parameters._
+. _Change the values of the tunable (structural) parameters. For tunable structural parameters, the "reconfiguration Mode" state must be entered before and left afterwards._
```
Comment (t-sommer) [inline docs/2_2_common_schema.adoc]: Are we using double quotes for states? If so this should go in the contribution guide.

File: docs/2_2_common_schema.adoc [hunk 6]
```diff
@@ -1032,14 +1081,14 @@ Even if this Dirac impulse would be modeled correctly by the modeling environmen
...
 
-. _Compute all parameters that depend on the tunable parameters._
+. _Compute all parameters (and sizes of variables, states, derivatives, zero crossings, ...) that depend on the tunable (structural) parameters._
```
Comment (t-sommer) [inline docs/2_2_common_schema.adoc]: Maybe we should stick with "event detection" vs. "zero crossing"?

File: docs/2_2_common_schema.adoc (hunk 1 above)
Comment (klausschuch) [reply docs/2_2_common_schema.adoc]: Just checked, for all other xml tags we don't use angle brackets in the schema description to refer to the tag, e.g.  we write `ModelVariables` instead of `<ModelVariables>`.
It might be a good idea to use angle brackets in the specification whenever an xml tag/element is referred.
For variables we have to consider that we should introduce a meta-tag `Variable`. For this meta-tag wie should not use the angle brackets but maybe curly brackets to not confuse it with an actual tag/element.

File: docs/2_2_common_schema.adoc (hunk 2 above)
Comment (klausschuch) [reply docs/2_2_common_schema.adoc]: I just fixed the typo here (two closing " instead of one "). This is part of a list of all allowed attribute values, where each value is written in "value".
I suggest that this pull request just contains the missing changes of the doc file.
This is a more general style issue.

File: docs/2_2_common_schema.adoc (hunk 3 above)
Comment (klausschuch) [reply docs/2_2_common_schema.adoc]: I missed the newly introduced inheritance for variables here.
Not sure what you mean with double quotes here. Can you change it like you think it should be done?
Would this be ok: `<{Variable} start="x" />`
(see discussion above about imaginary xml tag)

File: docs/2_2_common_schema.adoc (hunk 6 above)
Comment (klausschuch) [reply docs/2_2_common_schema.adoc]: right, it should be "event indicators"

File: docs/2_2_common_schema.adoc (hunk 5 above)
Comment (klausschuch) [reply docs/2_2_common_schema.adoc]: In other sections, we write the states without quotes but in bold letters. (I will fix it)

General topic:
We should try to avoid state names that consist of more than 1 word.
Additionally, we should either call it mode or state. Currently this is quite messy.

File: docs/2_2_common_schema.adoc (hunk 4 above)
Comment (klausschuch) [reply docs/2_2_common_schema.adoc]: I did this part before our decision to use plain english.
However, switching to plain english seems to be a bigger topic. (`=` for attribute values is used at several places). I think we should not mix up both changes. This pu request should only cover the content that was missing to cover arrays. There should be a general issue to harmonize equal vs `=` in the specification.

File: docs/2_2_common_schema.adoc (hunk 3 above)
Comment (t-sommer) [reply docs/2_2_common_schema.adoc]: Actually this should be `{VariableType}` as a placeholder for Float64 etc.
To me the curly braces look a bit confusion but that's a matter taste.
We should define the placeholders in a central place (including all possible values).

File: docs/2_2_common_schema.adoc (hunk 5 above)
Comment (klausschuch) [reply docs/2_2_common_schema.adoc]: Actually, my last comment is unfortunately not true: State names are not consistently written in bold in other sections. (sometimes bold, sometimes not)

I suggest to write state names bold.
